File: Chess/v3/Chess.py
import tkinter as tk
import random

TILE = 80

class AIChess:
    def __init__(self):
        self.root = tk.Tk()
        self.root.title("Chess vs Computer - Difficulty Options")
        self.canvas = tk.Canvas(self.root, width=640, height=740)
        self.canvas.pack()

        self.board = [
            ["bR","bN","bB","bQ","bK","bB","bN","bR"],
            ["bP","bP","bP","bP","bP","bP","bP","bP"],
            ["","","","","","","",""],
            ["","","","","","","",""],
            ["","","","","","","",""],
            ["","","","","","","",""],
            ["wP","wP","wP","wP","wP","wP","wP","wP"],
            ["wR","wN","wB","wQ","wK","wB","wN","wR"],
        ]

        self.selected = None
        self.white_turn = True
        self.vs_computer = True
        self.difficulty = 2
        self.message = "White to move | C toggle AI | 1/2/3 difficulty"

        self.canvas.bind("<Button-1>", self.click)
        self.root.bind("<KeyPress>", self.key)
        self.draw()
        self.root.mainloop()

    def diff_name(self):
        return ["", "Easy", "Medium", "Hard"][self.difficulty]

    def key(self, e):
        if e.keysym.lower() == "c":
            self.vs_computer = not self.vs_computer
            self.message = "Computer opponent ON" if self.vs_computer else "Two-player mode ON"
        elif e.keysym in ["1", "2", "3"]:
            self.difficulty = int(e.keysym)
            self.message = f"Difficulty set to {self.diff_name()}"
        self.draw()

    def draw_piece(self, piece, x, y):
        if not piece:
            return

        white = piece[0] == "w"
        kind = piece[1]
        main = "#f5f5eb" if white else "#232328"
        edge = "#777777" if white else "#e6e6e6"

        self.canvas.create_oval(x+18, y+58, x+62, y+70, fill="#000000", outline="")
        self.canvas.create_oval(x+20, y+48, x+60, y+64, fill=main, outline=edge, width=2)
        self.canvas.create_rectangle(x+25, y+28, x+55, y+58, fill=main, outline=edge, width=2)

        if kind == "P":
            self.canvas.create_oval(x+27, y+18, x+53, y+44, fill=main, outline=edge, width=2)
        elif kind == "R":
            self.canvas.create_rectangle(x+23, y+18, x+57, y+34, fill=main, outline=edge, width=2)
            for sx in [25, 37, 49]:
                self.canvas.create_rectangle(x+sx, y+12, x+sx+8, y+22, fill=main, outline=edge)
        elif kind == "N":
            pts = [x+25,y+52,x+31,y+20,x+43,y+12,x+56,y+25,x+50,y+31,x+58,y+48,x+45,y+54]
            self.canvas.create_polygon(pts, fill=main, outline=edge, width=2)
            self.canvas.create_oval(x+44, y+24, x+48, y+28, fill=edge, outline="")
        elif kind == "B":
            self.canvas.create_oval(x+24, y+12, x+56, y+47, fill=main, outline=edge, width=2)
            self.canvas.create_line(x+40, y+15, x+34, y+38, fill=edge, width=2)
        elif kind == "Q":
            self.canvas.create_oval(x+24, y+20, x+56, y+48, fill=main, outline=edge, width=2)
            for i in range(5):
                self.canvas.create_oval(x+20+i*9, y+9+(i%2)*3, x+30+i*9, y+19+(i%2)*3, fill=main, outline=edge)
        elif kind == "K":
            self.canvas.create_oval(x+24, y+20, x+56, y+48, fill=main, outline=edge, width=2)
            self.canvas.create_rectangle(x+37, y+8, x+43, y+28, fill=main, outline=edge)
            self.canvas.create_rectangle(x+30, y+14, x+50, y+19, fill=main, outline=edge)

    def draw(self):
        self.canvas.delete("all")
        for r in range(8):
            for c in range(8):
                color = "#eedab9" if (r+c)%2 == 0 else "#765230"
                self.canvas.create_rectangle(c*TILE, r*TILE, c*TILE+TILE, r*TILE+TILE, fill=color, outline="")
                if self.selected == (r, c):
                    self.canvas.create_rectangle(c*TILE+3, r*TILE+3, c*TILE+TILE-3, r*TILE+TILE-3, outline="yellow", width=5)
                self.draw_piece(self.board[r][c], c*TILE, r*TILE)

        self.canvas.create_rectangle(0,640,640,740, fill="#141414")
        self.canvas.create_text(320,675, text=self.message, fill="white", font=("Arial", 15, "bold"))
        mode = "vs Computer" if self.vs_computer else "2 Player"
        self.canvas.create_text(320,710, text=f"Mode: {mode} | Difficulty: {self.diff_name()}",
                                fill="white", font=("Arial", 16, "bold"))

    def click(self, e):
        if self.vs_computer and not self.white_turn:
            return

        row, col = e.y // TILE, e.x // TILE
        if not (0 <= row < 8 and 0 <= col < 8):
            return

        if self.selected is None:
            if self.board[row][col] and self.correct_turn(self.board[row][col]):
                self.selected = (row, col)
                self.message = "Selected " + self.board[row][col]
        else:
            sr, sc = self.selected
            if self.valid_move(self.board, sr, sc, row, col):
                self.move_piece(sr, sc, row, col)
                self.selected = None
                self.draw()
                if self.vs_computer and not self.white_turn:
                    self.message = "Computer thinking..."
                    self.draw()
                    self.root.after(350, self.computer_move)
                    return
            else:
                self.message = "Illegal move"
                self.selected = None
        self.draw()

    def correct_turn(self, piece):
        return (self.white_turn and piece[0] == "w") or ((not self.white_turn) and piece[0] == "b")

    def move_piece(self, sr, sc, er, ec):
        self.board[er][ec] = self.board[sr][sc]
        self.board[sr][sc] = ""
        self.white_turn = not self.white_turn
        self.message = "White to move" if self.white_turn else "Black to move"

    def all_moves(self, color):
        moves = []
        for sr in range(8):
            for sc in range(8):
                if self.board[sr][sc].startswith(color):
                    for er in range(8):
                        for ec in range(8):
                            if self.valid_move(self.board, sr, sc, er, ec):
                                moves.append((sr, sc, er, ec))
        return moves

    def computer_move(self):
        moves = self.all_moves("b")
        if not moves:
            self.message = "Computer has no legal moves"
            self.draw()
            return

        if self.difficulty == 1:
            move = random.choice(moves)
        elif self.difficulty == 2:
            move = self.greedy_move(moves)
        else:
            move = self.minimax_like_move(moves)

        self.move_piece(*move)
        self.message = "Computer moved | White to move"
        self.draw()

    def greedy_move(self, moves):
        best = moves[0]
        best_score = -999
        for move in moves:
            _, _, er, ec = move
            target = self.board[er][ec]
            score = self.value(target[1]) if target else 0
            score += random.randint(0, 2)
            if score > best_score:
                best_score = score
                best = move
        return best

    def minimax_like_move(self, moves):
        best = moves[0]
        best_score = -99999

        for move in moves:
            sr, sc, er, ec = move
            captured = self.board[er][ec]
            piece = self.board[sr][sc]
            self.board[er][ec] = piece
            self.board[sr][sc] = ""

            score = self.evaluate_board() - self.best_white_reply()

            self.board[sr][sc] = piece
            self.board[er][ec] = captured

            if score > best_score:
                best_score = score
                best = move

        return best

    def best_white_reply(self):
        best = 0
        for move in self.all_moves("w"):
            _, _, er, ec = move
            target = self.board[er][ec]
            if target:
                best = max(best, self.value(target[1]))
        return best

    def evaluate_board(self):
        score = 0
        for row in self.board:
            for p in row:
                if p:
                    v = self.value(p[1])
                    score += v if p[0] == "b" else -v
        return score

    def value(self, kind):
        return {"P":10, "N":30, "B":30, "R":50, "Q":90, "K":900}.get(kind, 0)

    def valid_move(self, b, sr, sc, er, ec):
        if (sr, sc) == (er, ec):
            return False
        piece = b[sr][sc]
        target = b[er][ec]
        if not piece or (target and target[0] == piece[0]):
            return False

        dr, dc = er-sr, ec-sc
        color, kind = piece[0], piece[1]

        if kind == "P":
            direction = -1 if color == "w" else 1
            start = 6 if color == "w" else 1
            if dc == 0 and dr == direction and not target: return True
            if dc == 0 and sr == start and dr == 2*direction and not target and not b[sr+direction][sc]: return True
            if abs(dc) == 1 and dr == direction and target: return True
        if kind == "R": return (dr == 0 or dc == 0) and self.clear(b, sr, sc, er, ec)
        if kind == "B": return abs(dr) == abs(dc) and self.clear(b, sr, sc, er, ec)
        if kind == "Q": return (dr == 0 or dc == 0 or abs(dr) == abs(dc)) and self.clear(b, sr, sc, er, ec)
        if kind == "K": return abs(dr) <= 1 and abs(dc) <= 1
        if kind == "N": return (abs(dr), abs(dc)) in [(2,1), (1,2)]
        return False

    def clear(self, b, sr, sc, er, ec):
        step_r = (er > sr) - (er < sr)
        step_c = (ec > sc) - (ec < sc)
        r, c = sr + step_r, sc + step_c
        while (r, c) != (er, ec):
            if b[r][c]:
                return False
            r += step_r
            c += step_c
        return True

AIChess()

File: Chess/v3/Chess.java
import java.awt.*;
import java.awt.event.*;
import java.util.*;
import javax.swing.*;

public class Chess extends JFrame {
    public Chess() {
        add(new ChessPanel());
        setTitle("Chess vs Computer - Difficulty Options");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setResizable(false);
        pack();
        setLocationRelativeTo(null);
        setVisible(true);
    }

    public static void main(String[] args) {
        new Chess();
    }
}

class ChessPanel extends JPanel implements MouseListener, KeyListener {
    private final int TILE = 80;
    private String[][] board = new String[8][8];
    private int selectedRow = -1, selectedCol = -1;
    private boolean whiteTurn = true;
    private boolean vsComputer = true;
    private int difficulty = 2; // 1 easy, 2 medium, 3 hard
    private String message = "White to move | C toggle AI | 1/2/3 difficulty";
    private Random rand = new Random();

    public ChessPanel() {
        setPreferredSize(new Dimension(640, 740));
        addMouseListener(this);
        addKeyListener(this);
        setFocusable(true);
        setupBoard();
    }

    private void setupBoard() {
        board[0] = new String[]{"bR","bN","bB","bQ","bK","bB","bN","bR"};
        board[1] = new String[]{"bP","bP","bP","bP","bP","bP","bP","bP"};
        for (int r = 2; r < 6; r++) for (int c = 0; c < 8; c++) board[r][c] = "";
        board[6] = new String[]{"wP","wP","wP","wP","wP","wP","wP","wP"};
        board[7] = new String[]{"wR","wN","wB","wQ","wK","wB","wN","wR"};
    }

    protected void paintComponent(Graphics g) {
        Graphics2D g2 = (Graphics2D)g;
        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        for (int r = 0; r < 8; r++) {
            for (int c = 0; c < 8; c++) {
                Color base = (r+c)%2==0 ? new Color(238,218,185) : new Color(118,82,48);
                g2.setPaint(new GradientPaint(c*TILE, r*TILE, base.brighter(), c*TILE+TILE, r*TILE+TILE, base.darker()));
                g2.fillRect(c*TILE, r*TILE, TILE, TILE);
                if (r == selectedRow && c == selectedCol) {
                    g2.setColor(new Color(255,230,0,160));
                    g2.fillRect(c*TILE, r*TILE, TILE, TILE);
                }
                drawPiece(g2, board[r][c], c*TILE, r*TILE);
            }
        }

        g2.setColor(new Color(20,20,20));
        g2.fillRect(0,640,640,100);
        g2.setColor(Color.WHITE);
        g2.setFont(new Font("Arial", Font.BOLD, 18));
        g2.drawString(message, 15, 675);
        g2.drawString("Mode: " + (vsComputer ? "vs Computer" : "2 Player") +
                " | Difficulty: " + diffName(), 15, 708);
    }

    private String diffName() {
        if (difficulty == 1) return "Easy";
        if (difficulty == 2) return "Medium";
        return "Hard";
    }

    private void drawPiece(Graphics2D g2, String piece, int x, int y) {
        if (piece.equals("")) return;
        boolean white = piece.charAt(0) == 'w';
        char type = piece.charAt(1);

        Color main = white ? new Color(245,245,235) : new Color(35,35,40);
        Color edge = white ? new Color(120,120,120) : new Color(230,230,230);

        g2.setColor(new Color(0,0,0,80));
        g2.fillOval(x+18,y+58,44,12);
        g2.setColor(main);
        g2.fillOval(x+20,y+48,40,16);
        g2.fillRoundRect(x+25,y+28,30,30,10,10);

        if (type == 'P') {
            g2.fillOval(x+27,y+18,26,26);
        } else if (type == 'R') {
            g2.fillRect(x+23,y+18,34,16);
            g2.fillRect(x+25,y+12,8,10);
            g2.fillRect(x+37,y+12,8,10);
            g2.fillRect(x+49,y+12,8,10);
        } else if (type == 'N') {
            Polygon horse = new Polygon();
            horse.addPoint(x+25,y+52); horse.addPoint(x+31,y+20); horse.addPoint(x+43,y+12);
            horse.addPoint(x+56,y+25); horse.addPoint(x+50,y+31); horse.addPoint(x+58,y+48);
            horse.addPoint(x+45,y+54);
            g2.fillPolygon(horse);
            g2.setColor(edge);
            g2.fillOval(x+44,y+24,4,4);
        } else if (type == 'B') {
            g2.fillOval(x+24,y+12,32,35);
            g2.setColor(edge);
            g2.drawLine(x+40,y+15,x+34,y+38);
        } else if (type == 'Q') {
            g2.fillOval(x+24,y+20,32,28);
            for (int i=0; i<5; i++) {
                g2.setColor(main);
                g2.fillOval(x+20+i*9,y+9+(i%2)*3,10,10);
            }
        } else if (type == 'K') {
            g2.fillOval(x+24,y+20,32,28);
            g2.fillRect(x+37,y+8,6,20);
            g2.fillRect(x+30,y+14,20,5);
        }

        g2.setColor(edge);
        g2.setStroke(new BasicStroke(2));
        g2.drawOval(x+20,y+48,40,16);
        g2.drawRoundRect(x+25,y+28,30,30,10,10);
    }

    public void mousePressed(MouseEvent e) {
        requestFocusInWindow();

        if (vsComputer && !whiteTurn) return;

        int col = e.getX() / TILE;
        int row = e.getY() / TILE;
        if (row < 0 || row >= 8 || col < 0 || col >= 8) return;

        if (selectedRow == -1) {
            if (!board[row][col].equals("") && correctTurn(board[row][col])) {
                selectedRow = row;
                selectedCol = col;
                message = "Selected " + board[row][col];
            }
        } else {
            if (validMove(board, selectedRow, selectedCol, row, col)) {
                movePiece(selectedRow, selectedCol, row, col);
                selectedRow = selectedCol = -1;

                if (vsComputer && !whiteTurn) {
                    message = "Computer thinking...";
                    repaint();
                    javax.swing.Timer t = new javax.swing.Timer(350, ev -> {
                        computerMove();
                        ((javax.swing.Timer)ev.getSource()).stop();
                    });
                    t.setRepeats(false);
                    t.start();
                }
            } else {
                message = "Illegal move";
                selectedRow = selectedCol = -1;
            }
        }
        repaint();
    }

    private void movePiece(int sr, int sc, int er, int ec) {
        board[er][ec] = board[sr][sc];
        board[sr][sc] = "";
        whiteTurn = !whiteTurn;
        message = whiteTurn ? "White to move" : "Black to move";
    }

    private boolean correctTurn(String p) {
        return whiteTurn && p.charAt(0)=='w' || !whiteTurn && p.charAt(0)=='b';
    }

    private ArrayList<Move> allMoves(String color) {
        ArrayList<Move> moves = new ArrayList<>();
        for (int sr=0; sr<8; sr++) {
            for (int sc=0; sc<8; sc++) {
                if (!board[sr][sc].equals("") && board[sr][sc].startsWith(color)) {
                    for (int er=0; er<8; er++) {
                        for (int ec=0; ec<8; ec++) {
                            if (validMove(board, sr, sc, er, ec)) {
                                moves.add(new Move(sr, sc, er, ec));
                            }
                        }
                    }
                }
            }
        }
        return moves;
    }

    private void computerMove() {
        ArrayList<Move> moves = allMoves("b");
        if (moves.isEmpty()) {
            message = "Computer has no legal moves";
            return;
        }

        Move best;
        if (difficulty == 1) {
            best = moves.get(rand.nextInt(moves.size()));
        } else if (difficulty == 2) {
            best = bestGreedyMove(moves);
        } else {
            best = bestMinimaxMove(moves);
        }

        movePiece(best.sr, best.sc, best.er, best.ec);
        message = "Computer moved | White to move";
        repaint();
    }

    private Move bestGreedyMove(ArrayList<Move> moves) {
        Move best = moves.get(0);
        int bestScore = -9999;
        for (Move m : moves) {
            String target = board[m.er][m.ec];
            int score = target.equals("") ? 0 : value(target.charAt(1));
            score += rand.nextInt(3);
            if (score > bestScore) {
                bestScore = score;
                best = m;
            }
        }
        return best;
    }

    private Move bestMinimaxMove(ArrayList<Move> moves) {
        Move best = moves.get(0);
        int bestScore = -999999;

        for (Move m : moves) {
            String captured = board[m.er][m.ec];
            String piece = board[m.sr][m.sc];

            board[m.er][m.ec] = piece;
            board[m.sr][m.sc] = "";

            int score = evaluateBoard() - bestWhiteReplyScore();

            board[m.sr][m.sc] = piece;
            board[m.er][m.ec] = captured;

            if (score > bestScore) {
                bestScore = score;
                best = m;
            }
        }

        return best;
    }

    private int bestWhiteReplyScore() {
        ArrayList<Move> replies = allMovesForBoard("w", board);
        int best = 0;
        for (Move m : replies) {
            String target = board[m.er][m.ec];
            int score = target.equals("") ? 0 : value(target.charAt(1));
            best = Math.max(best, score);
        }
        return best;
    }

    private ArrayList<Move> allMovesForBoard(String color, String[][] b) {
        ArrayList<Move> moves = new ArrayList<>();
        for (int sr=0; sr<8; sr++) {
            for (int sc=0; sc<8; sc++) {
                if (!b[sr][sc].equals("") && b[sr][sc].startsWith(color)) {
                    for (int er=0; er<8; er++) {
                        for (int ec=0; ec<8; ec++) {
                            if (validMove(b, sr, sc, er, ec)) moves.add(new Move(sr, sc, er, ec));
                        }
                    }
                }
            }
        }
        return moves;
    }

    private int evaluateBoard() {
        int score = 0;
        for (int r=0; r<8; r++) {
            for (int c=0; c<8; c++) {
                String p = board[r][c];
                if (!p.equals("")) {
                    int v = value(p.charAt(1));
                    score += p.charAt(0) == 'b' ? v : -v;
                }
            }
        }
        return score;
    }

    private int value(char type) {
        if (type == 'P') return 10;
        if (type == 'N' || type == 'B') return 30;
        if (type == 'R') return 50;
        if (type == 'Q') return 90;
        if (type == 'K') return 900;
        return 0;
    }

    private boolean validMove(String[][] b, int sr, int sc, int er, int ec) {
        if (sr == er && sc == ec) return false;
        String p = b[sr][sc], t = b[er][ec];
        if (p.equals("") || (!t.equals("") && t.charAt(0) == p.charAt(0))) return false;

        int dr = er - sr, dc = ec - sc;
        char color = p.charAt(0), type = p.charAt(1);

        if (type == 'P') {
            int dir = color == 'w' ? -1 : 1;
            int start = color == 'w' ? 6 : 1;
            if (dc == 0 && dr == dir && t.equals("")) return true;
            if (dc == 0 && sr == start && dr == 2*dir && t.equals("") && b[sr+dir][sc].equals("")) return true;
            if (Math.abs(dc) == 1 && dr == dir && !t.equals("")) return true;
        }
        if (type == 'R') return (dr == 0 || dc == 0) && clear(b, sr, sc, er, ec);
        if (type == 'B') return Math.abs(dr) == Math.abs(dc) && clear(b, sr, sc, er, ec);
        if (type == 'Q') return (dr == 0 || dc == 0 || Math.abs(dr) == Math.abs(dc)) && clear(b, sr, sc, er, ec);
        if (type == 'K') return Math.abs(dr) <= 1 && Math.abs(dc) <= 1;
        if (type == 'N') return Math.abs(dr) == 2 && Math.abs(dc) == 1 || Math.abs(dr) == 1 && Math.abs(dc) == 2;
        return false;
    }

    private boolean clear(String[][] b, int sr, int sc, int er, int ec) {
        int stepR = Integer.compare(er, sr), stepC = Integer.compare(ec, sc);
        int r = sr + stepR, c = sc + stepC;
        while (r != er || c != ec) {
            if (!b[r][c].equals("")) return false;
            r += stepR;
            c += stepC;
        }
        return true;
    }

    public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_C) {
            vsComputer = !vsComputer;
            message = vsComputer ? "Computer opponent ON" : "Two-player mode ON";
        }
        if (e.getKeyCode() == KeyEvent.VK_1) difficulty = 1;
        if (e.getKeyCode() == KeyEvent.VK_2) difficulty = 2;
        if (e.getKeyCode() == KeyEvent.VK_3) difficulty = 3;
        repaint();
    }

    public void mouseReleased(MouseEvent e) {}
    public void mouseClicked(MouseEvent e) {}
    public void mouseEntered(MouseEvent e) {}
    public void mouseExited(MouseEvent e) {}
    public void keyReleased(KeyEvent e) {}
    public void keyTyped(KeyEvent e) {}
}

class Move {
    int sr, sc, er, ec;
    Move(int sr, int sc, int er, int ec) {
        this.sr = sr; this.sc = sc; this.er = er; this.ec = ec;
    }
}
